src/plugins/explore/public/components/visualizations/chart_type_selector.tsx

```diff
@@ -0,0 +1,221 @@
+} from '@elastic/eui';
+import React, { useEffect, useMemo, useRef, useState } from 'react';
+import { useSelector } from 'react-redux';
+import { ALL_VISUALIZATION_RULES } from './rule_repository';
+import { VisualizationRule } from './types';
+import { ChartType, VisualizationTypeResult } from './utils/use_visualization_types';
+import { CHART_METADATA } from './constants';
+import { StyleAccordion } from './style_panel/style_accordion';
+import { selectChartType } from '../../application/utils/state_management/selectors';
+
+interface ChartTypeSelectorProps<T extends ChartType> {
+  visualizationData: VisualizationTypeResult<T>;
+  onChartTypeChange?: (chartType: ChartType) => void;
+}
+
+interface AvailableRuleOption {
+  value: string;
+  inputDisplay: React.ReactNode;
+  iconType?: string;
+  disabled?: boolean;
+  rules: Array<Partial<VisualizationRule>>;
+}
+
+export const ChartTypeSelector = <T extends ChartType>({
+  visualizationData,
+  onChartTypeChange,
+}: ChartTypeSelectorProps<T>) => {
+  // Indicates no rule is matched and the user should manually generate the visualization
+  const shouldManuallyGenerate = useRef(!Boolean(visualizationData.visualizationType));
+
+  // Local state for chart type, initialized from Redux
+  const storedChartTypeId = useSelector(selectChartType);
+  const [currChartTypeId, setCurrChartTypeId] = useState(storedChartTypeId);
+
+  useEffect(() => {
+    setCurrChartTypeId(storedChartTypeId);
+  }, [storedChartTypeId]);
+
+  // Map columns to add value and text fields for select component
+  const { numericalColumns = [], categoricalColumns = [], dateColumns = [] } = visualizationData;
+
+  // Get icon type based on chart type
+  const getChartIconType = (type: string): string => {
+    switch (type) {
+      case 'line':
+        return 'visLine';
+      case 'area':
+        return 'visArea';
+      case 'bar':
+        return 'visBarVertical';
+      case 'pie':
+        return 'visPie';
+      case 'metric':
+        return 'visMetric';
+      case 'heatmap':
+        return 'heatmap';
+      case 'scatterpoint':
+        return 'visScatter';
+      case 'table':
+        return 'tableOfContents';
+      default:
+        return 'visualizeApp';
+    }
+  };
+
+  // Create base mapping once with all visualization rules
+  const baseChartTypeMapping = useMemo(() => {
+    return ALL_VISUALIZATION_RULES.reduce((acc, rule) => {
+      rule.chartTypes.forEach(({ type, name }) => {
+        if (!acc[type]) {
+          const iconType = getChartIconType(type);
+          acc[type] = {
+            value: type,
+            inputDisplay: name,
+            iconType,
+            rules: [],
+          };
+        }
+        acc[type].rules.push({
+          id: rule.id,
+          name: rule.name,
+          matchIndex: rule.matchIndex,
+          toExpression: rule.toExpression,
+        });
+      });
+      return acc;
+    }, {} as Record<string, AvailableRuleOption>);
+  }, []); // Only calculate once
```

Comment:
Nit: Since this mapping only calculate once, I prefer to move this mapping and `getChartIconType` out of component body.